### PR TITLE
feat: add createAssertSnapshot

### LINKED ISSUE #2214: testing/snapshot: "assertSnapshot" should have a way to set options in a persistent way
**Is your feature request related to a problem? Please describe.**

Currently, if I want to set some options for each snapshot in a file, I need to pass these options to each call of `assertSnapshot`. For example:

```ts
Deno.test("example", async (t) => {
  assertSnapshot(t, "first", { dir: "__snaps__" });
  assertSnapshot(t, "second", { dir: "__snaps__" });
  assertSnapshot(t, "third", { dir: "__snaps__" });
  assertSnapshot(t, "fourth", { dir: "__snaps__" });
});
```

This makes tests harder to read.

**Describe the solution you'd like**

Ideally, it would be possible to do something like this:

```ts
assertSnapshot.setOptions({ dir: "__snaps__" });

Deno.test("example", async (t) => {
  assertSnapshot(t, "first");
  assertSnapshot(t, "second");
  assertSnapshot(t, "third");
  assertSnapshot(t, "fourth");
});
```

EDIT: On further consideration, I am actually leading slightly towards Option 2 below but would still be happy if this (or both) were implemented.

**Describe alternatives you've considered**

#### Option 1 - Wrapper Function

It is already possible to wrap assert snapshot like this:

```ts
async function assertSnapshotWithDefaultOpts<T>(
  t: Deno.TestContext,
  actual: T,
  options: SnapshotOptions<T>,
): Promise<void>;
export async function assertSnapshotWithDefaultOpts<T>(
  context: Deno.TestContext,
  actual: T,
  message?: string,
): Promise<void>;
export async function assertSnapshotWithDefaultOpts(
  context: Deno.TestContext,
  actual: unknown,
  msgOrOpts?: string | SnapshotOptions<unknown>,
): Promise<void> {
  await assertSnapshot(t, actual, {
    dir: "__snaps__",
    ...(typeof msgOrOpts === "object" ? msgOrOpts : { msg: msgOrOpts }),
  });
}

Deno.test("example", async (t) => {
  assertSnapshotWithDefaultOpts(t, "first");
  assertSnapshotWithDefaultOpts(t, "second");
  assertSnapshotWithDefaultOpts(t, "third");
  assertSnapshotWithDefaultOpts(t, "fourth");
});
```

However, this is unnecessarily verbose in my opinion.

#### Option 2 - Curried Function

A function (e.g. "createAssertSnapshot") could be exposed from `testing/snapshot.ts`.

```ts
const assertSnapshot = createAssertSnapshot({ /* options */ });

Deno.test("example", async (t) => {
  assertSnapshot(t, "first");
  assertSnapshot(t, "second");
  assertSnapshot(t, "third");
  assertSnapshot(t, "fourth");
});
```

This has the advantage that you could create different assertions and export them from a test utility file. For example,

```ts
export const assertSnapshotStr = createAssertSnapshot<string>({
  serializer: (actual: string) => /* --snip-- */,
});
export const assertSnapshotReactNode = createAssertSnapshot<React.ReactNode>({
  serializer: (actual: React.ReactNode) => /* --snip-- */,
  dir: "__components__",
});
```

### PR DESCRIPTION
**Pull request by bcheidemann**
Closes #2214

### CODE REVIEW

**Review by kt3k — COMMENTED** [review]
@bcheidemann Thanks for the PR. The API design and testing look good to me. Nice work!

Can you also add docs in `testing/README.md`?

File: testing/asserts.ts [hunk 1]
```diff
@@ -680,10 +680,10 @@ export function assertThrows<E extends Error = Error>(
 }
 
 /** Executes a function which returns a promise, expecting it to reject. */
-export function assertRejects(
+export function assertRejects<T = unknown>(
```
Comment (bartlomieju) [inline testing/asserts.ts:683]: Seems like an unrelated change, what's the reason for it?

**Review by bartlomieju — COMMENTED** [review]
Looking good!

File: testing/asserts.ts (hunk 1 above)
Comment (kt3k) [reply testing/asserts.ts:683]: Ah, probably it's better to keep this unknown as the thrown value is not type checked

File: testing/snapshot_test.ts [hunk 2]
```diff
@@ -491,3 +497,55 @@ Deno.test("Snapshot Test - Regression #2144", async (t) => {
...
+    );
+  });
+
+  await t.step("Message", async (t) => {
+    const assertMissingSnapshot = createAssertSnapshot<string>({
+      mode: "assert",
+      name: "[MISSING SNAPSHOT]",
+    });
+
+    const err = await assertRejects<AssertionError>(async () => {
```
Comment (kt3k) [inline testing/snapshot_test.ts:530]: Let's use `assertRejects(fn, AssertionError)` call, that will give correct type to `err`

File: testing/asserts.ts (hunk 1 above)
Comment (bcheidemann) [reply testing/asserts.ts:683]: Thanks for flagging, I will remove this change and refactor the test to use the format outlined below

**Comment by bcheidemann** [comment]
> @bcheidemann Thanks for the PR. The API design and testing look good to me. Nice work!
> 
> Can you also add docs in `testing/README.md`?

Yeah no problem :)

File: testing/README.md [hunk 3]
```diff
@@ -235,6 +235,71 @@ The `assertSnapshot` function will only attempt to read from and write to
 snapshot files. As such, the allow list for `--allow-read` and `--allow-write`
 can be limited to only include existing snapshot files, if so desired.
 
+### Options:
```
Comment (kt3k) [inline testing/README.md:238]: Nice docs!

**Review by kt3k — APPROVED** [review]
LGTM!
